openemr/openemr
# Billing process refactor and enhancements

@growlingflea and @kchapple have produced this this code which is aimed at improving the billing processing engine and adding some key enhancements.

Enhancements:
-Ability to configure x-12 partners so claims batch files are transmitted automatically using SFTP
-Ability to generate one x-12 batch-file per x-12 partner (Dan did this, which required a different x12 format)
-Billing processor refactoring so that each processing task is consolidated to a single file

Bugs fixed:
-date range broken on billing report query
-Logic errors on several lines in process_billing.php related to logical operator precedence $clear_claim
-Generate x-12 encounter" doesn't present any log or validation
-removed validate_payer_reset function, seems totally pointless.
-bill_log screen messages not printed when PDF download is initiated

bugs found, not addressed:
-provider_id is always zero, and should be set to rendering provider on the encounter
-updateClaim is not updating the last_level_billed because payer type is -1. Won't that interfere with multiple insurance billing?

For those who want to help test...

**Gen X-12 Regression test plan:**
1. check out latest master
2. load claim data into database. 
3. run generate-x12 in billing manager and save resulting batch file
4. check out the branch billing-processor-direct-x12-auto-sftp-rebase and run sql upgrade
5. re-run generate-x12 in billing manager ON THE SAME CLAIM(s) and save resulting batch file
6. compare the two batch files and if they're different, there may be a bug in the new batch file generator

**Testing new feature one batch-file per x-12 partner**
1. Enable the global Billing > "Generate X-12 Based On Insurance Company"
2. Fill out multiple x-12 partners and their SFTP settings in Practice Settings > x-12 partners
3. Run generate-x12 claims using various combinations of insurance companies and x-12 partners
4. Check the batch file results and make sure the contents reflect the settings correct partners and payors

**Testing new automate SFTP feature**
1. Enable the global Billing > "Automatically SFTP Claims To X12 Partner"
2. make sure you have valid and invalid SFTP credentials set for the SFTP settings in Practice Settings > x-12 partners
3. Run generate-x12 claims using various combinations of insurance companies and x-12 partners
4. Check SFTP results in Fees > Claim File Tracker and make sure the file is available there for download (if you ran "Continue" action.
5. Check "details" of batch rows which contain the patient ID and other claim data.
6. Check your valid remote SFTP server to make sure the claim was successfully transmitted (if status == 'success')
